core/src/Properties/Kokkos_Detection.hpp

```diff
@@ -0,0 +1,173 @@
+          class... Args>
+struct detector_workaround_impl {
+  // This could be done with inheritance, but let's just do it this way
+  // to preempt some mistakes that could be made involving that and that
+  // wouldn't work with the "real" detection idiom
+  using _impl                 = WorkaroundImpl<Default, void, Args...>;
+  using type                  = typename _impl::type;
+  static constexpr auto value = _impl::value;
+};
+
+}  // end namespace _detection_impl
+
+#define KOKKOS_DECLARE_DETECTION_ARCHETYPE(name, params, params_no_intro, ...) \
+  template <class Default, class _k__always_void,                              \
+            KOKKOS_PP_REMOVE_PARENS(params)>                                   \
+  struct name##__impl {                                                        \
+    static constexpr bool value = false;                                       \
+    using type                  = Default;                                     \
+  };                                                                           \
+  template <class Default, KOKKOS_PP_REMOVE_PARENS(params)>                    \
+  struct name##__impl<Default, ::Kokkos::Impl::void_t<__VA_ARGS__>,            \
+                      KOKKOS_PP_REMOVE_PARENS(params_no_intro)> {              \
+    static constexpr bool value = true;                                        \
+    using type                  = __VA_ARGS__;                                 \
+  };                                                                           \
+  template <class Default, class _always_void,                                 \
+            KOKKOS_PP_REMOVE_PARENS(params)>                                   \
+  using name = name##__impl<Default, _always_void,                             \
+                            KOKKOS_PP_REMOVE_PARENS(params_no_intro)>
+
+template <class Default, class _always_void, template <class...> class Op,
+          class... Args>
+using _detector =
+    _detection_impl::detector_workaround_impl<Default, Op, Args...>;
+
+#endif
+
+#define KOKKOS_DECLARE_DETECTION_ARCHETYPE_1PARAM(name, param, ...) \
+  KOKKOS_DECLARE_DETECTION_ARCHETYPE(name, (class param), (param), __VA_ARGS__)
+
+#define KOKKOS_DECLARE_DETECTION_ARCHETYPE_2PARAMS(name, param1, param2, ...) \
+  KOKKOS_DECLARE_DETECTION_ARCHETYPE(name, (class param1, class param2),      \
+                                     (param1, param2), __VA_ARGS__)
+
+    template <template <class...> class Op, class... Args>
+    using is_detected = _detector<nonesuch, void, Op, Args...>;
+
+template <template <class...> class Op, class... Args>
+using detected_t = typename is_detected<Op, Args...>::type;
+
+template <class Default, template <class...> class Op, class... Args>
+using detected_or = _detector<Default, void, Op, Args...>;
+
+template <class Default, template <class...> class Op, class... Args>
+using detected_or_t = typename detected_or<Default, Op, Args...>::type;
```

Comment:
Not used anywhere.

Reply:
This is just the standard detection idiom stuff.  I thought it would be more recognizable if I didn't remove stuff from Walter Brown's original implementation.

Reply:
Please add a reference to the original implementation.

Reply:
This has been done